src/openfl/_internal/renderer/canvas/CanvasBitmap.hx

```diff
@@ -13,18 +13,20 @@ class CanvasBitmap {
 
 		if (bitmap.__bitmapData != null && bitmap.__bitmapData.__isValid && bitmap.__bitmapData.__canBeDrawnToCanvas()) {
 			renderSession.blendModeManager.setBlendMode(bitmap.__worldBlendMode);
-			renderSession.maskManager.pushObject(bitmap, false);
 
-			if (!renderSession.allowSmoothing || !bitmap.smoothing) {
-				CanvasSmoothing.setEnabled(context, false);
-			}
+			var needRender = renderSession.maskManager.pushObject(bitmap, false);
+			if (needRender) {
+				if (!renderSession.allowSmoothing || !bitmap.smoothing) {
+					CanvasSmoothing.setEnabled(context, false);
+				}
 
-			context.globalAlpha = bitmap.__worldAlpha;
-			bitmap.__bitmapData.__drawToCanvas(context, bitmap.__renderTransform, renderSession.roundPixels || bitmap.__snapToPixel(),
-				renderSession.pixelRatio, bitmap.__scrollRect, true);
+				context.globalAlpha = bitmap.__worldAlpha;
+				bitmap.__bitmapData.__drawToCanvas(context, bitmap.__renderTransform, renderSession.roundPixels || bitmap.__snapToPixel(),
+					renderSession.pixelRatio, bitmap.__scrollRect, true);
 
-			if (!renderSession.allowSmoothing || !bitmap.smoothing) {
-				CanvasSmoothing.setEnabled(context, true);
+				if (!renderSession.allowSmoothing || !bitmap.smoothing) {
+					CanvasSmoothing.setEnabled(context, true);
+				}
 			}
 
 			renderSession.maskManager.popObject(bitmap, false);
```

Comment:
Always move the `popObject` into the `if (needRender)` to avoid restoring the context, when it actually was never saved?

Reply:
You're right! Just moving `popObject` won't work because there might still be a rect needs popping, but something has to be done here anyway. Maybe an (inlined) `withMasks(() -> actualRender())` would be nicer, although these days I prefer more dumb/explicit code for low-level layers like this.